app/services/sign_in/token_exchanger.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module SignIn
+  class TokenExchanger
+    attr_reader :subject_token, :subject_token_type, :actor_token, :actor_token_type, :client_id
+
+    def initialize(subject_token:, subject_token_type:, actor_token:, actor_token_type:, client_id:)
+      @subject_token = subject_token
+      @subject_token_type = subject_token_type
+      @actor_token = actor_token
+      @actor_token_type = actor_token_type
+      @client_id = client_id
+    end
+
+    def perform
+      validations
+      new_container = create_new_session
+      Rails.logger.info("[SignIn][TokenExchanger] success, new_session_handle=#{new_container.session.handle}")
+      new_container
+    end
+
+    private
+
+    def validations
+      validate_subject_token
+      validate_subject_token_type
+      validate_actor_token
+      validate_actor_token_type
+      validate_client_id
+      validate_device_sso
+    end
+
+    def validate_subject_token
+      unless subject_token && current_access_token
+        raise Errors::InvalidTokenError.new message: 'subject token is invalid'
+      end
+    end
+
+    def validate_subject_token_type
+      unless subject_token_type == Constants::Urn::ACCESS_TOKEN
+        raise Errors::InvalidTokenTypeError.new message: 'subject token type is invalid'
+      end
+    end
+
+    def validate_actor_token_type
+      unless actor_token_type == Constants::Urn::DEVICE_SECRET
+        raise Errors::InvalidTokenTypeError.new message: 'actor token type is invalid'
+      end
+    end
+
+    def validate_actor_token
+      hashed_actor_token = hash_token(actor_token)
+
+      unless hashed_actor_token == current_session.hashed_device_secret &&
+             hashed_actor_token == current_access_token.device_secret_hash
+        raise Errors::InvalidTokenError.new message: 'actor token is invalid'
+      end
+    end
+
+    def hash_token(token)
+      Digest::SHA256.hexdigest(token)
+    end
+
+    def validate_client_id
+      unless new_session_client_config&.shared_sessions
```

Comment:
🤔 The usage of the safe navigation operator makes me wonder if we shouldn't throw an error if the client config isn't found in `new_session_client_config`.

Reply:
👍